### DISCUSSION TITLE
Fixtures in multiple conftest files run in nested order

### DISCUSSION THREAD

**Discussion by guy-saban-ws** [pytest-dev/pytest#8475] [Q&A]
I have the following project structure

    ├─Root
    ├── Folder
    └── Tests
        ├── conftest.py
        ├── TestsA
        │    ├── SubFolderAA
        │    │    ├── test_aaa.py
        │    │    └── test_aab.py
        │    └── SubFolderAB
        │         ├── test_aba.py
        │         └── test_abb.py
        └─── TestsB
             ├── SubFolderBA
             │    ├── test_baa.py
             │    └── test_bab.py
             └── SubFolderBB
                  ├── test_bba.py
                  └── test_bbb.py

I have a fixture I want to run for every Tests_ folder (`TestsA` and `TestsB`), so I created a fixture with scope package

    @pytest.fixture(scope='package', autouse=True)
    def package_setup_and_teardown():
        # setup
        yield
        # teardown

When added in `conftest.py` under Tests it runs only ones although I added `__init__.py` file to `TestsA` and `TestsB` to make them a package.

When I added a `conftest.py` with the fixture to each of the Tests_ folders both executed the setup code before the tests under it, but executed the teardown code only after all the tests finished, and as if they were executed as nested:

    package_setup_and_teardown for TestsA (setup)
    tests under TestsA
    package_setup_and_teardown for TestsB (setup)
    tests under TestsB
    package_setup_and_teardown for TestsB (teardown)
    package_setup_and_teardown for TestsA (teardown)

Which causes some issues, especially since I have more than two Tests_ folders.

Why does it behave like this, and more important how can I solve it without flattening the test hierarchy or adding `conftest.py` to every SubFolder.

I'm using Python 3.8 and Pytest 6.2.2 if it makes any difference.

**Comment by portyu9** [comment]
Hi @guy-saban-ws 

`scope="package"` is tied to pytest's package/collection boundaries; one package-scoped fixture defined above several sibling packages is not a request for "run once independently for every descendant directory."

If `TestsA` and `TestsB` are intended to be separate packages, make them actual package boundaries and expose the fixture from each package's `conftest.py`:

```text
Tests/
    TestsA/
        __init__.py
        conftest.py
    TestsB/
        __init__.py
        conftest.py
```

You do not need to duplicate the fixture implementation. Put the implementation in a shared plugin/module, and import/re-export it from each package's `conftest.py`.

That gives pytest an unambiguous package fixture boundary for each sibling.

The posted result was also produced with pytest 6.2.2; collection/package handling has changed substantially since then. I would upgrade before building a workaround around that old nested teardown behavior.

Please mark this answer as accepted if it helped, thank you.
